File: crm/app/internal/adapters/inbound/messenger/messenger_parser.go
// Package messenger — inbound adapter that parses FB Messenger webhook payloads.
//
// Scope (v1): parse payload JSON → []domain.ParsedMessage.
// Live FB Graph API fetch is OUT OF SCOPE for v1 (no token on this host).
//
// TODO (live integration): To enable real-time ingest, implement a transport layer
// that either:
//   a) Receives the FB webhook POST (verify token + X-Hub-Signature-256 check), or
//   b) Polls FB Graph API: GET /{page-id}/conversations?fields=messages{...}&access_token=TOKEN
// Both paths produce the same raw JSON that ParseWebhookPayload already handles.
// Wire the transport to call ParseWebhookPayload, then ConversationService.IngestMessage
// for each returned ParsedMessage. The token/transport plumbing belongs in cmd/server
// or a dedicated crm/sync/ingest_messenger.go (Python or Go) — not here.
//
// Hexagonal position: inbound adapter (no imports of sqlite or application packages).
package messenger

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// WebhookPayload is the top-level FB Messenger webhook body.
// FB sends: {"object":"page","entry":[{"id":"<page_id>","messaging":[...]}]}
type WebhookPayload struct {
	Object string        `json:"object"`
	Entry  []EntryObject `json:"entry"`
}

// EntryObject represents one page entry in the webhook.
type EntryObject struct {
	ID        string            `json:"id"` // FB page id
	Messaging []MessagingObject `json:"messaging"`
}

// MessagingObject represents one send/receive event.
type MessagingObject struct {
	Sender    ParticipantRef `json:"sender"`
	Recipient ParticipantRef `json:"recipient"`
	Timestamp int64          `json:"timestamp"` // Unix ms
	Message   *MessageEvent  `json:"message,omitempty"`
}

// ParticipantRef holds an PSID or page id.
type ParticipantRef struct {
	ID string `json:"id"`
}

// MessageEvent is the nested message object within a MessagingObject.
type MessageEvent struct {
	MID         string       `json:"mid"`
	Text        string       `json:"text,omitempty"`
	IsEcho      bool         `json:"is_echo,omitempty"`
	Attachments []Attachment `json:"attachments,omitempty"`
}

// Attachment is a FB Messenger attachment (image, video, file, etc.).
type Attachment struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

// ParseWebhookPayload parses a raw FB Messenger webhook JSON body and returns
// one domain.ParsedMessage per message event (inbound and echo/outbound).
//
// Only message events (messaging[].message != nil) are processed; delivery/read
// receipts and other event types are silently skipped.
//
// Thread key: always keyed on the customer psid. For echo events (page-sent),
// sender.id is the page_id and the customer psid is in recipient.id.
// Direction: "in" for customer→page, "out" for page→customer (echo).
//
// Timestamps: events with ms <= 0 are skipped with a logged reason rather than
// using time.Now() as a fallback, which would produce incorrect data.
func ParseWebhookPayload(raw []byte) ([]domain.ParsedMessage, error) {
	var payload WebhookPayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, fmt.Errorf("messenger parser: unmarshal payload: %w", err)
	}

	var out []domain.ParsedMessage
	for _, entry := range payload.Entry {
		pageID := entry.ID
		for _, ev := range entry.Messaging {
			if ev.Message == nil {
				continue // skip delivery/read receipts
			}
			mid := ev.Message.MID
			if mid == "" {
				continue // malformed; skip
			}

			// Reject bad timestamps rather than silently using time.Now().
			sentAt, err := msToUTC(ev.Timestamp)
			if err != nil {
				fmt.Printf("messenger parser: skipping mid=%s: %v\n", mid, err)
				continue
			}

			// Determine direction and thread key.
			// Echo events have sender.id == pageID (or is_echo flag set).
			// The conversation thread is always keyed on the CUSTOMER psid.
			isEcho := ev.Message.IsEcho || ev.Sender.ID == pageID
			direction := "in"
			threadPSID := ev.Sender.ID
			senderRef := ev.Sender.ID
			if isEcho {
				direction = "out"
				// For page-sent/echo, the customer psid is in recipient.id.
				threadPSID = ev.Recipient.ID
				senderRef = ev.Sender.ID // keep actual sender (page id)
			}

			// Serialise attachments as JSON array string (empty string when none).
			attachments := ""
			if len(ev.Message.Attachments) > 0 {
				b, _ := json.Marshal(ev.Message.Attachments)
				attachments = string(b)
			}

			out = append(out, domain.ParsedMessage{
				Channel:           "messenger",
				ExternalThreadID:  threadPSID, // always the customer psid
				PageID:            pageID,
				ExternalMessageID: mid,
				Direction:         direction,
				SenderRef:         senderRef,
				Body:              ev.Message.Text,
				Attachments:       attachments,
				SentAt:            sentAt,
			})
		}
	}
	return out, nil
}

// msToUTC converts a Unix millisecond timestamp to UTC ISO-8601 string.
// Returns an error for ms <= 0 rather than falling back to time.Now(),
// so callers can skip malformed events instead of recording wrong timestamps.
func msToUTC(ms int64) (string, error) {
	if ms <= 0 {
		return "", fmt.Errorf("invalid timestamp ms=%d (must be > 0)", ms)
	}
	return time.UnixMilli(ms).UTC().Format("2006-01-02T15:04:05.000Z"), nil
}

File: crm/app/application/messenger_parser.py
"""Inbound adapter: parse FB Messenger webhook payloads → list[ParsedMessage].

Scope (v1): payload JSON → normalised messages.
Live FB Graph API fetch is out of scope (no token on this host).

TODO (live integration): wire a transport layer that either:
  a) Receives the FB webhook POST (verify token + X-Hub-Signature-256), or
  b) Polls GET /{page-id}/conversations?fields=messages{...}&access_token=TOKEN
Both paths produce the same JSON that parse_messenger_webhook already handles.
"""
from __future__ import annotations

import json
import logging
from datetime import datetime, timezone

from domain.entities.conversation import ParsedMessage

log = logging.getLogger(__name__)


def parse_messenger_webhook(payload: dict) -> list[ParsedMessage]:
    """Extract messages from a Facebook Messenger webhook payload.

    Processes message events only; delivery/read receipts and other event
    types are silently skipped. Skips events with invalid timestamps rather
    than falling back to now() and producing incorrect data.

    Thread key is always the customer PSID:
    - Inbound (customer→page): sender.id is the customer PSID.
    - Echo/outbound (page→customer): sender.id is the page; recipient.id is the customer PSID.

    Returns a flat list of ParsedMessage, one per valid message event.
    """
    out: list[ParsedMessage] = []

    for entry in payload.get("entry", []):
        page_id: str = entry.get("id", "")
        for ev in entry.get("messaging", []):
            message = ev.get("message")
            if message is None:
                continue  # delivery/read receipt — skip

            mid: str = message.get("mid", "")
            if not mid:
                continue  # malformed — skip

            ts_ms: int = ev.get("timestamp", 0)
            sent_at = _ms_to_utc(ts_ms)
            if sent_at is None:
                log.warning("messenger parser: skipping mid=%s: invalid timestamp ms=%s", mid, ts_ms)
                continue

            sender_id: str = ev.get("sender", {}).get("id", "")
            recipient_id: str = ev.get("recipient", {}).get("id", "")

            # Echo when is_echo flag is set OR sender is the page itself.
            is_echo: bool = bool(message.get("is_echo")) or sender_id == page_id
            if is_echo:
                direction = "out"
                thread_psid = recipient_id   # customer psid
                sender_ref = sender_id       # page id (actual sender)
            else:
                direction = "in"
                thread_psid = sender_id      # customer psid
                sender_ref = sender_id

            # Serialise attachments as JSON string; empty string when none.
            raw_attachments = message.get("attachments") or []
            attachments = json.dumps(raw_attachments) if raw_attachments else ""

            out.append(ParsedMessage(
                channel="messenger",
                external_thread_id=thread_psid,
                page_id=page_id,
                external_message_id=mid,
                direction=direction,
                sender_ref=sender_ref,
                body=message.get("text", ""),
                attachments=attachments,
                sent_at=sent_at,
            ))

    return out


def _ms_to_utc(ms: int) -> str | None:
    """Convert Unix millisecond timestamp to UTC ISO-8601 string.

    Returns None for ms <= 0 so callers can skip malformed events instead
    of recording wrong timestamps.
    """
    if ms <= 0:
        return None
    dt = datetime.fromtimestamp(ms / 1000.0, tz=timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"
